### DISCUSSION TITLE
Current Status and Roadmap

### DISCUSSION THREAD

**Discussion by jacomyal** [jacomyal/sigma.js#1469] [Announcements]
# Sigma.js: Current Status and Roadmap

Hey everyone,

This post is here to explain where sigma.js is at right now, what’s coming in v3, and what to expect for v4.

### Current State of Sigma.js

Sigma.js v3 has been used in many projects for a while now, and overall, things are looking good. For example, we’re using it in [Gephi Lite](https://github.com/gephi/gephi-lite) at @ouestware, and @arthur-bigeard-gdotv (who sponsored the developments of the v3) has been using it heavily in his software [G.V()](https://gdotv.com/). I’m pretty confident in the quality of the new renderers code (which was the big change in v3). The only reason the `beta` tag is still there is that we never set a plan to remove it.

That said, there are still some tricky issues we haven’t fully solved yet:
- [Opacity management](https://github.com/jacomyal/sigma.js/issues/1426)
- [Depth management](https://github.com/jacomyal/sigma.js/issues/1427)
- [Labels management](https://github.com/jacomyal/sigma.js/issues/1319)

Right now, I don’t see an easy way to fix these without breaking things. So to keep things clear, I’ve set two new milestones.

### Planning for v3 Release

The [v3 milestone](https://github.com/jacomyal/sigma.js/milestone/10) is mostly about small features that could help some users, as well as fixing the touch support.

I hope we’ll be able to wrap this up and remove the `beta` tag before the end of the year.

### Looking Far Ahead to v4

The [v4 milestone](https://github.com/jacomyal/sigma.js/milestone/11) is for bigger, breaking changes. This will focus on rewriting the renderers API again with things like [Order-independent Transparency (OIT)](https://en.wikipedia.org/wiki/Order-independent_transparency), doing picking and rendering in the same step, and rendering edges and nodes in the same layer. 

We’ll probably also move the labels API from Canvas to DOM + CSS, to make it easier to customize labels and allow interactions with them.

We’re planning to start this in 2025, but there’s no fixed schedule yet.

### Moving Forward

We at @ouestware are really happy with where sigma.js is right now. A big thank you again to Arthur from [G.V()](https://gdotv.com/) for sponsoring this version! His support allowed us to clean up the codebase, make sigma.js more stable, and improve the documentation.

Now that things are in better shape, it’ll be easier to maintain the project and fix any bugs that come up. 

While we’re not planning to break anything with major updates, new features will still be added regularly as separate packages. For example, we have plans for a cleaner and improved PNG exporter and a self-loops renderer.

Thanks again for your continued support!

**Comment by rahst12** [comment]
The progress v3 has made is great!

Definitely looking forward to v4 to see this issue resolved: https://github.com/jacomyal/sigma.js/issues/1427

Noted on issue #1427, we're interested to see if clustering/grouping of nodes will be supported, multi-node selection, and the ability to collapse/expand sections of the graph.

**Reply by jacomyal** [reply]
The two features you are expecting will probably not be developed directly as features of sigma. It is the same issue with nodes dragging: Sigma does not modify the graph state, it only reads it to render an image.

But it does not mean it is not possible to implement:

- Multi-node selections is very feasible right now, but it has to be done higher-level. It is implemented in [Gephi-Lite](https://github.com/gephi/gephi-lite), for instance:

[sigma-multi-selection.webm](https://github.com/user-attachments/assets/87467bd4-8823-4981-9c9e-e30c80bfd3c5)

- I have not implemented collapse/expand in any application yet, but you can implement it using [`animateNodes`](https://www.sigmajs.org/docs/typedoc/sigma/src/utils/functions/animateNodes). You can check the [Layouts example](https://www.sigmajs.org/storybook/?path=/story/layouts--story) story, to see how to use that helper

**Comment by jacomyal** [comment]
Quick update: Sigma finally hit v3.0.0 :tada:

**Comment by rahst12** [comment]
I've been following the activity in GitHub here.  It's great to see how stable v3 has been since its release at the start of the year.  With the state of v3, I'm curious if there's been any thought given to a roadmap update for v4, or just in general.

Thanks for all the hard work!

**Reply by jacomyal** [reply]
Hi @rahst12,

Short-term wise, there are still a few features I'd love to find the time to implement as satellite packages (self-loops for instance, rotate with right-click, lasso selection...), but currently our (@ouestware) open-source efforts are more focused on [Gephi Lite](https://github.com/gephi/gephi-lite).

About the v4, I hope to get a better view on how to tackle the core issue I have, i.e. [order-independant transparency](https://en.wikipedia.org/wiki/Order-independent_transparency) after the summer, but right now it still kind of scares me ^^

**Comment by rahst12** [comment]
I saw some activity on the v4 branch.  Any thoughts for the community on those?  Thanks!

**Reply by jacomyal** [reply]
Hi @rahst12,

I'm close to a beta release, I'll communicate about that v4 then here, in a new discussion.
I hope that it happens next week. Stay tuned :)

**Reply by jacomyal** [reply]
It's definitely taken longer than I expected, but it's here:
https://github.com/jacomyal/sigma.js/discussions/1539
